Modules/IP/Sources/IP/IP.swift

```diff
@@ -18,56 +27,347 @@ public enum IPMode {
+                try body($0)
+            }
+        }
+    }
+
+    public let address: Address
+
+    public var port: Int {
+        return address.port
+    }
+
+    public init(address: Address) {
         self.address = address
     }
 
-    public init(port: Int = 0, mode: IPMode = .ipV4Prefered) throws {
-        try IP.assertValid(port)
-        let address = iplocal(nil, Int32(port), mode.code)
-        try ensureLastOperationSucceeded()
+    public var veniceAddress: ipaddr {
+        var source = address
+        var veniceAddress = ipaddr()
+        withUnsafePointer(to: &source) { source in
+            withUnsafeMutablePointer(to: &veniceAddress) { veniceAddress in
+                memcpy(veniceAddress, source, MemoryLayout<ipaddr>.size)
+                return
+            }
+        }
+        return veniceAddress
+    }
+
+    public init(veniceAddress: ipaddr) {
+        var source = veniceAddress
+        var address = Address()
+        withUnsafePointer(to: &source) { source in
+            withUnsafeMutablePointer(to: &address) { destination in
+                memcpy(destination, source, MemoryLayout<Address>.size)
+                return
+            }
+        }
         self.init(address: address)
     }
 
-    public init(localAddress: String, port: Int = 0, mode: IPMode = .ipV4Prefered) throws {
-        try IP.assertValid(port)
-        let address = iplocal(localAddress, Int32(port), mode.code)
-        try ensureLastOperationSucceeded()
+    public init(port: Int, mode: IPMode = .ipV4Prefered) throws {
+        let address = try IP.ip(port: port, mode: mode)
         self.init(address: address)
     }
 
-    public init(remoteAddress: String, port: Int, mode: IPMode = .ipV4Prefered, deadline: Double = .never) throws {
-        try IP.assertValid(port)
-        let address = ipremote(remoteAddress, Int32(port), mode.code, deadline.int64milliseconds)
-        try ensureLastOperationSucceeded()
+    public init(localAddress address: String, port: Int = 0, mode: IPMode = .ipV4Prefered) throws {
+        let address = try IP.ip(address: address, port: port, mode: mode)
         self.init(address: address)
     }
 
-    private static func assertValid(_ port: Int) throws {
+    // TODO:
+    // public init(interfaceName: String, port: Int = 0, mode: IPMode = .ipV4Prefered) throws {}
+
+    public init(remoteAddress address: String, port: Int, mode: IPMode = .ipV4Prefered, deadline: Double = .never) throws {
+        let address = try IP.ip(address: address, port: port, mode: mode, deadline: deadline)
+        self.init(address: address)
+    }
+}
+
+extension IP {
+    private static func assertValidPort(_ port: Int) throws {
         if port < 0 || port > 0xffff {
             throw IPError.invalidPort
         }
     }
+
+    fileprivate static func ip(address: String, port: Int, mode: IPMode, deadline: Double) throws -> Address {
```

Comment:
maybe this should be named `address` instead of `ip`? `let address = IP.address(port:)` makes more sense than `let address = IP.ip(port:)`.